Events

fitzgen opened the pull request.

fitzgen committed:
Add event handling infrastructure

fitzgen committed:
Ensure that we always encode opcodes and immediates as u32s w/out padding

fitzgen committed:
Port examples/hello-world to the no-modules target

fitzgen committed:
Stop exposing ChangeDiscriminant to JS

We're not using it from JS and it is just adding bloat to the JS glue.

fitzgen committed:
Port examples/counter to the no-modules target

fitzgen committed:
Port examples/input-form to the no-modules target

fitzgen committed:
Update build script for no-modules examples

fitzgen committed:
Upgrade bumpalo

No more BumpAllocSafe

fitzgen committed:
Stop patching crates.io crates

Now that new versions are published.

fitzgen committed:
Fix doc tests

fitzgen commented:
r? @alexcrichton -- could you take a look at the second commit in particular? thanks!

alexcrichton commented on Cargo.toml:
```diff
@@ -46,3 +51,12 @@ members = [
   "./examples/input-form/crate",
 ]
+
+[patch.crates-io]
```
I find the large array of patches here quite amusing, although arguably a bug in Cargo so perhaps less so...

alexcrichton commented on examples/counter/crate/src/lib.rs:
```diff
@@ -28,59 +28,53 @@ impl Render for Counter {
+                Node::element(
+                    bump,
+                    "button",
+                    [on(bump, "click", |root, vdom, _event| {
```
Would it be possible to do something like `bump.on("click", ...)`? Maybe via something like a newtype wrapper or the like?

fitzgen replied:
I was thinking of adding an extension trait, and also thinking of adding `bump.div(...)` as shorthand for `Node::element(bump, "div", ...)`. Was wondering if that was a little too magicky or not though. Also, just generally not too focused on ergo yet, and hoping I can mostly kick that down the road to something like the `typed-html` crate.

alexcrichton commented on examples/counter/crate/src/lib.rs:
```diff
@@ -28,59 +28,53 @@ impl Render for Counter {
+                    bump,
+                    "button",
+                    [on(bump, "click", |root, vdom, _event| {
+                        root.unwrap_mut::<Counter>().decrement();
```
So just to confirm again from before, how would "middlewary style" libraries end up handling this? They don't know the type of the root object (right?), or would they have their own root vdom objects they can manipulate?

fitzgen replied:
They would be parameterized by some trait and would translate raw web events into "semantic" events for that component, and these semantic events would have a corresponding trait method.

I have TodoMVC implemented in the next PR after this, and it does this with its single `Todo` component, which you can sneak a peak at here: https://github.com/fitzgen/dodrio/blob/fill-out-examples/examples/todomvc/src/todo.rs#L9-L42

alexcrichton commented on examples/counter/crate/src/lib.rs:
```diff
@@ -28,59 +28,53 @@ impl Render for Counter {
+                    "button",
+                    [on(bump, "click", |root, vdom, _event| {
+                        root.unwrap_mut::<Counter>().decrement();
+                        vdom.schedule_render();
```
Also to confirm from before, what's the use case for *not* scheduling a rerender on a handled event?

fitzgen replied:
If the event only sends a request to the server and doesn't change the UI, or you detect that a re-render is unnecessary for whatever other reason.

It is the most general thing, since we don't expose any "unschedule render" methods (and I don't think that would make sense).

alexcrichton commented on examples/counter/crate/src/lib.rs:
```diff
@@ -28,59 +28,53 @@ impl Render for Counter {
-            [increment, Node::text(val.into_bump_str()), decrement],
+            [],
+            [
+                Node::element(
```
Was the movement from locals to this array stylistic or required for lifetimes?

fitzgen replied:
Stylistic.

alexcrichton commented on examples/input-form/crate/src/lib.rs:
```diff
@@ -26,6 +26,18 @@ impl Render for SayHelloTo {
+                    .and_then(|t| t.dyn_into::<web_sys::HtmlInputElement>().ok())
+                {
+                    Some(input) => input,
+                    None => return,
```
This may have just answered my previous question about when you wouldn't want to schedule a render

fitzgen replied:
;)

alexcrichton commented on js/change-list.js:
```diff
@@ -175,8 +250,14 @@ class ChangeList {
       // console.log(OP_TABLE[op].name);
       // console.log(this.stack);
-      i = OP_TABLE[op](this.stack, mem8, mem32, i);
+      i = OP_TABLE[op](this.stack, this.eventsTrampoline, this.listeners, mem8, mem32, i);
```
Random stylistic question, should the table take `this` instead of the fields of `this`?

fitzgen replied:
Perhaps! I just kept adding one more thing at a time... By now, it likely should just take `this`!

It is also *possible* that passing the pieces is faster if the `this` properties don't end up monomorphic in the JIT's eyes, but that (a) seems unlikely given, and (b) shouldn't matter relative to the DOM method costs / is definitely premature optimization.

alexcrichton commented on src/vdom.rs:
```diff
@@ -48,76 +153,114 @@ where
+            .into_inner()
+            .component
+            .take()
+            .unwrap_throw()
```
Now this is a doozy of a chain of methods

fitzgen replied:
...yes

alexcrichton commented on src/vdom.rs:
```diff
@@ -143,39 +286,73 @@ where
+        debug!("  updating event listeners");
+
+        'outer1: for new_l in new {
+            unsafe {
```
So I've ended up here sort of randomly without actually knowing much context for what's going on, but as a random stylistic note might be good to have a comment about the unsafety here (even if it just points elsewhere)

fitzgen replied:
Good call. The comment you were looking for is on the `EventsRegistry::add` method, FWIW.

alexcrichton commented on src/vdom.rs:
```diff
@@ -258,25 +440,191 @@ where
+}
+
+impl<T> CellOptionExt<T> for Cell<Option<T>> {
+    fn take(&self) -> Option<T> {
```
I think `cell.take()` is the same as `cell.replace(None)`, right?

fitzgen replied:
...yes

good catch

fitzgen replied:
And somehow it compiles just fine without this extension trait too -- so there is a `take` method for `Cell<Option<T>>` somewhere already??

fitzgen replied:
https://doc.rust-lang.org/nightly/std/cell/struct.Cell.html#method.take

fitzgen replied:
how the heck did I miss this when reading over these methods??

alexcrichton replied:
bahahaha

alexcrichton reviewed (commented):
So I definitely won't pretend to have fully understood all that just happened, but overall looks great to me! At least a solid base to build on :)

fitzgen commented:
Thanks for giving this a look!

fitzgen committed:
Remove compile-tests since they are more pain than their worth

Very hard to get them to actually run correctly with anything but a clean target
dir, which makes them very slow and annoying to run. Not worth it for a single
test.

fitzgen committed:
Make change list opcode implementations take a `ChangeList` instead of all of its parts

Kept adding one more field, and all of a sudden we have almost all the
fields. Easier to just pass a reference to the whole `ChangeList`.

fitzgen committed:
Remove commented out console.log calls

fitzgen committed:
Remove redundant extension trait

It turns out that `Cell<T>` already has a `take` method that leaves the default
in its place. And for `Option<T>` the default is `None`, so it does exactly what
we want already.

fitzgen committed:
Fill out missing doc comment

fitzgen committed:
Add a comment about `unsafe` operation
